Goal of this notebook:

Investigate the impact of replacing NA values in the dataset with a placeholder on the classification accuracy with decision tree and random forest.  (XGBoost Random Forest works with NA values natively)

In [1]:
import numpy as np
import pandas as pd
from IPython.core.display import display

pd.set_option('display.max_columns', None)
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
import warnings
import classifier_utils
warnings.filterwarnings("ignore", category=UserWarning)

In [2]:
non_features_columns = ["chunk_id", "line_start", "line_end", "line_separator", "kind_conflict", "url", "project"]
non_features_columns.extend(["project_user", "project_name", "path", "file_name", "sha", "leftsha", "rightsha", "basesha"])

In [3]:
selected_dataset = pd.read_csv("../../data/SELECTED_LABELLED_DATASET.csv")
projects = list(selected_dataset['project'].unique())

# Decision Tree

In [4]:
dt = DecisionTreeClassifier(random_state=99)

In [5]:
result_dt = classifier_utils.ProjectsResults(dt, projects, non_features_columns)

In [6]:
report_dt = result_dt.get_report_df(include_overall=True)
report_dt

project  observations  \
0                                     CCI-MIT__XCoLab          5512   
1                            apache__directory-server           845   
2                                    jgralab__jgralab          2072   
3                                    Unidata__thredds          1154   
4               CloudStack-extras__CloudStack-archive          1424   
5                                    apache__accumulo          4113   
6                              Ramblurr__Anki-Android           892   
7                                     getrailo__railo           815   
8                              TeamDev-Ltd__OpenFaces          2979   
9                                 apache__lucene-solr          1256   
10                                          zkoss__zk          1087   
11                     AOKP__frameworks_base_disabled          1999   
12                   cgjones__android-frameworks-base           828   
13                                      freenet__fred          1268   
14                           hibernate__hibernate-orm          1000   
15                                       alexo__wro4j          1663   
16            allwinner-ics__platform_frameworks_base          1780   
17                             eucalyptus__eucalyptus          1020   
18                  android__platform_frameworks_base          3557   
19  RealVNC__android-frameworks-base-with-screensh...          2269   
20                                    atlasapi__atlas          1107   
21           IceColdSandwich__android_frameworks_base          1792   
22                              sebastianbenz__Jnario          8298   
23                              alkacon__opencms-core           923   
24                               SINTEF-9012__ThingML          1022   
25                             elastic__elasticsearch          1246   
26                             eclipse__jetty.project          1537   
27                        revolsys__com.revolsys.open           991   
28  adetaylor__android-frameworks-base-with-remote...          1792   
29                                            Overall         56241   

    observations (wt NaN)  precision   recall  f1-score  accuracy  \
0                    3757    0.97200  0.97200   0.97200   0.97200   
1                     652    0.92900  0.93100   0.93000   0.93100   
2                    1802    0.83500  0.83600   0.83500   0.83600   
3                     950    0.90700  0.90900   0.90800   0.90900   
4                    1106    0.75900  0.76000   0.75900   0.76000   
5                    3148    0.83400  0.83400   0.83400   0.83400   
6                     759    0.69000  0.69400   0.69200   0.69400   
7                     572    0.64800  0.64300   0.64600   0.64300   
8                    2859    0.96500  0.96500   0.96500   0.96500   
9                     974    0.55700  0.55900   0.55800   0.55900   
10                    881    0.76000  0.75600   0.75800   0.75600   
11                   1370    0.80500  0.80700   0.80600   0.80700   
12                    565    0.83300  0.82800   0.83000   0.82800   
13                   1012    0.59600  0.59400   0.59400   0.59400   
14                    716    0.50700  0.50700   0.50600   0.50700   
15                   1368    0.49700  0.49900   0.49800   0.49900   
16                   1187    0.79100  0.78500   0.78800   0.78500   
17                    747    0.56900  0.56500   0.56700   0.56500   
18                   2460    0.77500  0.77800   0.77600   0.77800   
19                   1561    0.77600  0.77700   0.77700   0.77700   
20                    782    0.56900  0.56800   0.56800   0.56800   
21                   1193    0.78800  0.78500   0.78700   0.78500   
22                   8173    0.99800  0.99900   0.99800   0.99900   
23                    840    0.94200  0.94200   0.94200   0.94200   
24                    977    0.93400  0.93300   0.93300   0.93300   
25                      0        NaN      NaN     

In [7]:
import importlib
importlib.reload(classifier_utils)
result_dt2 = classifier_utils.ProjectsResults(dt, projects, non_features_columns, replace_na=True)

In [8]:
report_dt2 = result_dt2.get_report_df(include_overall=True)
report_dt2

project  observations  \
0                                     CCI-MIT__XCoLab          5512   
1                            apache__directory-server           845   
2                                    jgralab__jgralab          2072   
3               CloudStack-extras__CloudStack-archive          1424   
4                                    Unidata__thredds          1154   
5                                    apache__accumulo          4113   
6                                     getrailo__railo           815   
7                              Ramblurr__Anki-Android           892   
8                              TeamDev-Ltd__OpenFaces          2979   
9                                 apache__lucene-solr          1256   
10                                      freenet__fred          1268   
11                                          zkoss__zk          1087   
12                             elastic__elasticsearch          1246   
13                   cgjones__android-frameworks-base           828   
14                                       alexo__wro4j          1663   
15                           hibernate__hibernate-orm          1000   
16  adetaylor__android-frameworks-base-with-remote...          1792   
17                                    atlasapi__atlas          1107   
18                     AOKP__frameworks_base_disabled          1999   
19                             eclipse__jetty.project          1537   
20  RealVNC__android-frameworks-base-with-screensh...          2269   
21                             eucalyptus__eucalyptus          1020   
22                  android__platform_frameworks_base          3557   
23            allwinner-ics__platform_frameworks_base          1780   
24           IceColdSandwich__android_frameworks_base          1792   
25                        revolsys__com.revolsys.open           991   
26                              alkacon__opencms-core           923   
27                               SINTEF-9012__ThingML          1022   
28                              sebastianbenz__Jnario          8298   
29                                            Overall         56241   

    observations (wt NaN)  precision    recall  f1-score  accuracy  \
0                    5512      0.972  0.971000  0.971000  0.971000   
1                     845      0.906  0.907000  0.906000  0.907000   
2                    2072      0.819  0.819000  0.819000  0.819000   
3                    1424      0.781  0.787000  0.784000  0.787000   
4                    1154      0.896  0.896000  0.896000  0.896000   
5                    4113      0.848  0.848000  0.848000  0.848000   
6                     815      0.727  0.730000  0.728000  0.730000   
7                     892      0.695  0.680000  0.687000  0.680000   
8                    2979      0.954  0.955000  0.955000  0.955000   
9                    1256      0.541  0.540000  0.540000  0.540000   
10                   1268      0.625  0.623000  0.624000  0.623000   
11                   1087      0.740  0.739000  0.739000  0.739000   
12                   1246      0.529  0.527000  0.528000  0.527000   
13                    828      0.838  0.835000  0.835000  0.835000   
14                   1663      0.498  0.497000  0.498000  0.497000   
15                   1000      0.495  0.494000  0.494000  0.494000   
16                   1792      0.815  0.815000  0.815000  0.815000   
17                   1107      0.587  0.584000  0.585000  0.584000   
18                   1999      0.802  0.802000  0.802000  0.802000   
19                   1537      0.563  0.561000  0.562000  0.561000   
20                   2269      0.791  0.792000  0.791000  0.792000   
21                   1020      0.565  0.560000  0.562000  0.560000   
22                   3557      0.792  0.789000  0.791000  0.789000   
23                   1780      0.796  0.793000  0.794000  0.793000   
24                   1792      0.800  0.799000  0.799000  0.799000   
25                    99

In [9]:
df_inner_dt = pd.merge(report_dt, report_dt2, on='project', how='inner', suffixes=('_withoutNA', '_withNA'))

accuracy_dt = df_inner_dt.filter(regex=("project|accuracy.*")).copy()
accuracy_dt['improvement'] = accuracy_dt.apply(lambda x: classifier_utils.get_normalized_improvement(x['accuracy_withNA'], x['accuracy_withoutNA']), axis=1)
accuracy_dt

project  accuracy_withoutNA  \
0                                     CCI-MIT__XCoLab             0.97200   
1                            apache__directory-server             0.93100   
2                                    jgralab__jgralab             0.83600   
3                                    Unidata__thredds             0.90900   
4               CloudStack-extras__CloudStack-archive             0.76000   
5                                    apache__accumulo             0.83400   
6                              Ramblurr__Anki-Android             0.69400   
7                                     getrailo__railo             0.64300   
8                              TeamDev-Ltd__OpenFaces             0.96500   
9                                 apache__lucene-solr             0.55900   
10                                          zkoss__zk             0.75600   
11                     AOKP__frameworks_base_disabled             0.80700   
12                   cgjones__android-frameworks-base             0.82800   
13                                      freenet__fred             0.59400   
14                           hibernate__hibernate-orm             0.50700   
15                                       alexo__wro4j             0.49900   
16            allwinner-ics__platform_frameworks_base             0.78500   
17                             eucalyptus__eucalyptus             0.56500   
18                  android__platform_frameworks_base             0.77800   
19  RealVNC__android-frameworks-base-with-screensh...             0.77700   
20                                    atlasapi__atlas             0.56800   
21           IceColdSandwich__android_frameworks_base             0.78500   
22                              sebastianbenz__Jnario             0.99900   
23                              alkacon__opencms-core             0.94200   
24                               SINTEF-9012__ThingML             0.93300   
25                             elastic__elasticsearch                 NaN   
26                             eclipse__jetty.project                 NaN   
27                        revolsys__com.revolsys.open                 NaN   
28  adetaylor__android-frameworks-base-with-remote...                 NaN   
29                                            Overall             0.76904   

    accuracy_withNA  improvement  
0          0.971000    -0.001029  
1          0.907000    -0.025779  
2          0.819000    -0.020335  
3          0.896000    -0.014301  
4          0.787000     0.112500  
5          0.848000     0.084337  
6          0.680000    -0.020173  
7          0.730000     0.243697  
8          0.955000    -0.010363  
9          0.540000    -0.033989  
10         0.739000    -0.022487  
11         0.802000    -0.006196  
12         0.835000     0.040698  
13         0.623000     0.071429  
14         0.494000    -0.025641  
15         0.497000    -0.004008  
16         0.793000     0.037209  
17         0.560000    -0.008850  
18         0.789000     0.049550  
19         0.792000     0.067265  
20         0.584000     0.037037  
21         0.799000     0.065116  
22         0.998000    -0.001001  
23         0.927000    -0.015924  
24         0.925000    -0.008574  
25         0.527000          NaN  
26         0.561000          NaN  
27         0.792000          NaN  
28         0.815000          NaN  
29         0.758103    -0.014221

The overall average accuracy with decision tree decreased from 0.76904 to 0.758103. It decreased in 15 of 25 projects and increased in 10.
On the other hand, inputing a constant to replace the NA values allowed to go from 25 to 29 projects.

# Random forest

In [10]:
model_3 = RandomForestClassifier(random_state=99, n_jobs=5, n_estimators=400, max_features=0.3, min_samples_leaf=1)# Decision Tree

In [11]:
result_dt3 = classifier_utils.ProjectsResults(model_3, projects, non_features_columns, replace_na=False)
report_dt3 = result_dt3.get_report_df(include_overall=True)

In [12]:
report_dt3

project  observations  \
0                                     CCI-MIT__XCoLab          5512   
1                            apache__directory-server           845   
2                                    jgralab__jgralab          2072   
3               CloudStack-extras__CloudStack-archive          1424   
4                                    Unidata__thredds          1154   
5                                    apache__accumulo          4113   
6                              Ramblurr__Anki-Android           892   
7                                     getrailo__railo           815   
8                                 apache__lucene-solr          1256   
9                              TeamDev-Ltd__OpenFaces          2979   
10                                          zkoss__zk          1087   
11                   cgjones__android-frameworks-base           828   
12                                      freenet__fred          1268   
13                           hibernate__hibernate-orm          1000   
14                                       alexo__wro4j          1663   
15                                    atlasapi__atlas          1107   
16                     AOKP__frameworks_base_disabled          1999   
17                  android__platform_frameworks_base          3557   
18            allwinner-ics__platform_frameworks_base          1780   
19  RealVNC__android-frameworks-base-with-screensh...          2269   
20                             eucalyptus__eucalyptus          1020   
21                              alkacon__opencms-core           923   
22           IceColdSandwich__android_frameworks_base          1792   
23                               SINTEF-9012__ThingML          1022   
24                              sebastianbenz__Jnario          8298   
25                             elastic__elasticsearch          1246   
26                             eclipse__jetty.project          1537   
27                        revolsys__com.revolsys.open           991   
28  adetaylor__android-frameworks-base-with-remote...          1792   
29                                            Overall         56241   

    observations (wt NaN)  precision   recall  f1-score  accuracy  \
0                    3757      0.971  0.97600   0.97300   0.97600   
1                     652      0.935  0.93900   0.93600   0.93900   
2                    1802      0.867  0.86900   0.86700   0.86900   
3                    1106      0.798  0.80300   0.79800   0.80300   
4                     950      0.913  0.91900   0.91400   0.91900   
5                    3148      0.860  0.86400   0.86000   0.86400   
6                     759      0.718  0.74000   0.72200   0.74000   
7                     572      0.687  0.70100   0.69300   0.70100   
8                     974      0.636  0.64400   0.63800   0.64400   
9                    2859      0.967  0.96900   0.96800   0.96900   
10                    881      0.767  0.78300   0.77300   0.78300   
11                    565      0.854  0.86200   0.85600   0.86200   
12                   1012      0.667  0.67400   0.66900   0.67400   
13                    716      0.591  0.60200   0.58800   0.60200   
14                   1368      0.573  0.58400   0.57500   0.58400   
15                    782      0.652  0.66600   0.65400   0.66600   
16                   1370      0.812  0.82400   0.81700   0.82400   
17                   2460      0.811  0.81800   0.81300   0.81800   
18                   1187      0.799  0.81300   0.80400   0.81300   
19                   1561      0.801  0.81000   0.80400   0.81000   
20                    747      0.598  0.62000   0.60600   0.62000   
21                    840      0.947  0.96200   0.95500   0.96200   
22                   1193      0.792  0.81100   0.79900   0.81100   
23                    977      0.936  0.94900   0.94200   0.94900   
24                   8173      0.998  0.99900   0.99900   0.99900   
25                      0        NaN      NaN     

In [13]:
result_dt4 = classifier_utils.ProjectsResults(model_3, projects, non_features_columns, replace_na=True)
report_dt4 = result_dt4.get_report_df(include_overall=True)

In [14]:
report_dt4

project  observations  \
0                                     CCI-MIT__XCoLab          5512   
1                            apache__directory-server           845   
2                                    jgralab__jgralab          2072   
3               CloudStack-extras__CloudStack-archive          1424   
4                                    Unidata__thredds          1154   
5                                    apache__accumulo          4113   
6                                     getrailo__railo           815   
7                              Ramblurr__Anki-Android           892   
8                                 apache__lucene-solr          1256   
9                                           zkoss__zk          1087   
10                             TeamDev-Ltd__OpenFaces          2979   
11                           hibernate__hibernate-orm          1000   
12                                      freenet__fred          1268   
13                   cgjones__android-frameworks-base           828   
14                             elastic__elasticsearch          1246   
15                                    atlasapi__atlas          1107   
16                                       alexo__wro4j          1663   
17                     AOKP__frameworks_base_disabled          1999   
18                             eclipse__jetty.project          1537   
19            allwinner-ics__platform_frameworks_base          1780   
20                  android__platform_frameworks_base          3557   
21                             eucalyptus__eucalyptus          1020   
22  adetaylor__android-frameworks-base-with-remote...          1792   
23  RealVNC__android-frameworks-base-with-screensh...          2269   
24           IceColdSandwich__android_frameworks_base          1792   
25                              alkacon__opencms-core           923   
26                               SINTEF-9012__ThingML          1022   
27                        revolsys__com.revolsys.open           991   
28                              sebastianbenz__Jnario          8298   
29                                            Overall         56241   

    observations (wt NaN)  precision   recall  f1-score  accuracy  \
0                    5512   0.973000  0.97700  0.974000   0.97700   
1                     845   0.930000  0.93300  0.931000   0.93300   
2                    2072   0.865000  0.86700  0.864000   0.86700   
3                    1424   0.814000  0.82000  0.815000   0.82000   
4                    1154   0.907000  0.91300  0.908000   0.91300   
5                    4113   0.872000  0.87900  0.874000   0.87900   
6                     815   0.756000  0.77400  0.763000   0.77400   
7                     892   0.741000  0.76300  0.745000   0.76300   
8                    1256   0.637000  0.64200  0.633000   0.64200   
9                    1087   0.786000  0.80400  0.792000   0.80400   
10                   2979   0.957000  0.96100  0.958000   0.96100   
11                   1000   0.602000  0.61700  0.603000   0.61700   
12                   1268   0.669000  0.67800  0.671000   0.67800   
13                    828   0.858000  0.87000  0.863000   0.87000   
14                   1246   0.618000  0.61100  0.611000   0.61100   
15                   1107   0.669000  0.68800  0.670000   0.68800   
16                   1663   0.592000  0.60100  0.593000   0.60100   
17                   1999   0.829000  0.84600  0.835000   0.84600   
18                   1537   0.639000  0.65600  0.635000   0.65600   
19                   1780   0.821000  0.83800  0.828000   0.83800   
20                   3557   0.823000  0.83400  0.826000   0.83400   
21                   1020   0.627000  0.65000  0.634000   0.65000   
22                   1792   0.823000  0.83900  0.828000   0.83900   
23                   2269   0.811000  0.82900  0.817000   0.82900   
24                   1792   0.808000  0.82900  0.816000   0.82900   
25                    923   0.928000  0.94600  0.9

In [15]:
df_inner_rf = pd.merge(report_dt3, report_dt4, on='project', how='inner', suffixes=('_withoutNA', '_withNA'))

accuracy_rf = df_inner_rf.filter(regex=("project|accuracy.*")).copy()
accuracy_rf['improvement'] = accuracy_rf.apply(lambda x: classifier_utils.get_normalized_improvement(x['accuracy_withNA'], x['accuracy_withoutNA']), axis=1)
accuracy_rf

project  accuracy_withoutNA  \
0                                     CCI-MIT__XCoLab             0.97600   
1                            apache__directory-server             0.93900   
2                                    jgralab__jgralab             0.86900   
3               CloudStack-extras__CloudStack-archive             0.80300   
4                                    Unidata__thredds             0.91900   
5                                    apache__accumulo             0.86400   
6                              Ramblurr__Anki-Android             0.74000   
7                                     getrailo__railo             0.70100   
8                                 apache__lucene-solr             0.64400   
9                              TeamDev-Ltd__OpenFaces             0.96900   
10                                          zkoss__zk             0.78300   
11                   cgjones__android-frameworks-base             0.86200   
12                                      freenet__fred             0.67400   
13                           hibernate__hibernate-orm             0.60200   
14                                       alexo__wro4j             0.58400   
15                                    atlasapi__atlas             0.66600   
16                     AOKP__frameworks_base_disabled             0.82400   
17                  android__platform_frameworks_base             0.81800   
18            allwinner-ics__platform_frameworks_base             0.81300   
19  RealVNC__android-frameworks-base-with-screensh...             0.81000   
20                             eucalyptus__eucalyptus             0.62000   
21                              alkacon__opencms-core             0.96200   
22           IceColdSandwich__android_frameworks_base             0.81100   
23                               SINTEF-9012__ThingML             0.94900   
24                              sebastianbenz__Jnario             0.99900   
25                             elastic__elasticsearch                 NaN   
26                             eclipse__jetty.project                 NaN   
27                        revolsys__com.revolsys.open                 NaN   
28  adetaylor__android-frameworks-base-with-remote...                 NaN   
29                                            Overall             0.80804   

    accuracy_withNA  improvement  
0           0.97700     0.041667  
1           0.93300    -0.006390  
2           0.86700    -0.002301  
3           0.82000     0.086294  
4           0.91300    -0.006529  
5           0.87900     0.110294  
6           0.76300     0.088462  
7           0.77400     0.244147  
8           0.64200    -0.003106  
9           0.96100    -0.008256  
10          0.80400     0.096774  
11          0.87000     0.057971  
12          0.67800     0.012270  
13          0.61700     0.037688  
14          0.60100     0.040865  
15          0.68800     0.065868  
16          0.84600     0.125000  
17          0.83400     0.087912  
18          0.83800     0.133690  
19          0.82900     0.100000  
20          0.65000     0.078947  
21          0.94600    -0.016632  
22          0.82900     0.095238  
23          0.94300    -0.006322  
24          0.99900     0.000000  
25          0.61100          NaN  
26          0.65600          NaN  
27          0.80500          NaN  
28          0.83900          NaN  
29          0.80731    -0.000903

The overall average accuracy with random forest decreased from 0.80804 to 0.80731. It decreased in 8 of 25 projects, increased in 16 and remained constant in 1.
On the other hand, inputing a constant to replace the NA values allowed to go from 25 to 29 projects.